ead/edhoc-ead-zeroconf/src/lib.rs

```diff
@@ -68,7 +121,159 @@ pub fn i_prepare_ead_3() -> Option<EADItem> {
+) -> EdhocMessageBuffer {
+    let (k_1, iv_1) = compute_k_1_iv_1(&prk);
+
+    // plaintext = (ID_U: bstr)
+    let mut plaintext = EdhocMessageBuffer::new();
+    plaintext.content[0] = CBOR_MAJOR_BYTE_STRING + id_u.len as u8;
+    plaintext.content[1..1 + id_u.len].copy_from_slice(&id_u.content[..id_u.len]);
+    plaintext.len = 1 + id_u.len;
+
+    // external_aad = (SS: int)
+    let enc_structure = encode_enc_structure(ss);
+
+    // ENC_ID = 'ciphertext' of COSE_Encrypt0
+    aes_ccm_encrypt_tag_8(&k_1, &iv_1, &enc_structure[..], &plaintext)
+}
+
+fn compute_prk(a: &BytesP256ElemLen, g_b: &BytesP256ElemLen) -> BytesHashLen {
+    // NOTE: salt should be h'' (the zero-length byte string), but crypto backends are hardcoded to salts of size SHA256_DIGEST_LEN (32).
+    //       using a larger but all-zeroes salt seems to generate the same result though.
+    let salt: BytesHashLen = [0u8; SHA256_DIGEST_LEN];
+    let g_ab = p256_ecdh(a, g_b);
+    hkdf_extract(&salt, &g_ab)
+}
+
+fn compute_k_1_iv_1(prk: &BytesHashLen) -> (BytesCcmKeyLen, BytesCcmIvLen) {
+    // K_1 = EDHOC-Expand(PRK, info = (0, h'', AES_CCM_KEY_LEN), length)
+    let mut k_1: BytesCcmKeyLen = [0x00; AES_CCM_KEY_LEN];
+    let k_1_buf = edhoc_kdf(prk, 0, &[0x00; MAX_KDF_CONTEXT_LEN], 0, AES_CCM_KEY_LEN);
+    k_1[..].copy_from_slice(&k_1_buf[..AES_CCM_KEY_LEN]);
+
+    // IV_1 = EDHOC-Expand(PRK, info = (1, h'', AES_CCM_KEY_LEN), length)
+    let mut iv_1: BytesCcmIvLen = [0x00; AES_CCM_IV_LEN];
+    // NOTE (FIXME?): here we actually generate AES_CCM_KEY_LEN bytes, but then we only use AES_CCM_IV_LEN of them (next line)
+    let iv_1_buf = edhoc_kdf(prk, 1, &[0x00; MAX_KDF_CONTEXT_LEN], 0, AES_CCM_KEY_LEN);
+    iv_1[..].copy_from_slice(&iv_1_buf[..AES_CCM_IV_LEN]);
+
+    (k_1, iv_1)
+}
+
+const EAD_ENC_STRUCTURE_LEN: usize = 2 + 8 + 3;
+fn encode_enc_structure(ss: u8) -> [u8; EAD_ENC_STRUCTURE_LEN] {
+    let mut encrypt0: Bytes8 = [0x00; 8];
+    encrypt0[0] = 0x45u8; // 'E'
+    encrypt0[1] = 0x6eu8; // 'n'
+    encrypt0[2] = 0x63u8; // 'c'
+    encrypt0[3] = 0x72u8; // 'r'
+    encrypt0[4] = 0x79u8; // 'y'
+    encrypt0[5] = 0x70u8; // 'p'
+    encrypt0[6] = 0x74u8; // 't'
+    encrypt0[7] = 0x30u8; // '0'
+
+    let mut enc_structure: [u8; EAD_ENC_STRUCTURE_LEN] = [0x00; EAD_ENC_STRUCTURE_LEN];
+
+    // encode Enc_structure from rfc9052 Section 5.3
+    enc_structure[0] = CBOR_MAJOR_ARRAY | 3 as u8; // 3 is the fixed number of elements in the array
+    enc_structure[1] = CBOR_MAJOR_TEXT_STRING | encrypt0.len() as u8;
+    enc_structure[2..2 + encrypt0.len()].copy_from_slice(&encrypt0[..]);
+    enc_structure[encrypt0.len() + 2] = CBOR_MAJOR_BYTE_STRING | 0x00 as u8; // 0 for zero-length byte string (empty Header)
+    enc_structure[encrypt0.len() + 3] = CBOR_MAJOR_BYTE_STRING | 0x01 as u8; // 1 for the `ss` value
+    enc_structure[encrypt0.len() + 4] = ss;
+
+    enc_structure
+}
+
+// NOTE: can we import this from the edhoc-rs main crate?
+fn edhoc_kdf(
```

Comment:
For the time being I moved just the logic for the `info` struct to a `helpers` module in `consts`.

Part of the reason is to avoid having `consts` depend on the cryptographic backends.
Another part is that this is now only a single small duplicated function, so in my view it still does not justify changing the dependencies in `consts`.
Nevertheless I added a TO-DO (🙃), that we should address as more cases like this arise.